### PR TITLE
Load balancing tool mods

### PR DESCRIPTION
**Pull request by jedwards4b**
A rework of load balancing tool which is better integrated with CIME and uses PFS test.

Test suite: hand testing tools/load_balancing_tool/test/test_load_balancing_tool.py
Test baseline: 
Test namelist changes: 
Test status: bit for bit

Fixes 

User interface changes?: 

Update gh-pages html (Y/N)?: Y

Code review:

### CODE REVIEW

**Comment by rljacob** [comment]
The right thing to do here is accept 1902 and then accept this one.

File: tools/load_balancing_tool/layouts.py [hunk 1]
```diff
@@ -154,3 +154,174 @@ def write_pe_file(self, pefilename):
             else:
                 nthrds[c] = 1
         self.write_pe_template(pefilename, ntasks, nthrds, roots)
+import optimize_model
```
Comment (jedwards4b) [inline tools/load_balancing_tool/layouts.py]: these are redundent - I will remove

File: tools/load_balancing_tool/optimize_model.py [hunk 2]
```diff
@@ -317,8 +313,8 @@ def get_state(self):
     def get_state_string(self, state):
         return self.states[state]
 
-    def get_all_variables(self):
-        return self.prob.X
+#    def get_all_variables(self):
```
Comment (jedwards4b) [inline tools/load_balancing_tool/optimize_model.py]: will remove dead code.

**Comment by sarich** [comment]
I'm getting errors on submitting now:
    Case dir: /homes/sarich/acme/scratch/PFS_I1.f19_g16.X.test_lbt_
    Errors were:
        Test 'PFS_I1.f19_g16.X' failed in phase 'CREATE_NEWCASE' with exception 'ERROR: PFS_I1.f19_g16.X is not a full test name'

Do we need to add <machine-name>_<compiler> to the test name?

**Comment by sarich** [comment]
OK, I see machine and compiler should be added by _create_newcase_phase, but that's not working for me.

**Comment by sarich** [comment]
This is failing, in test_scheduler.py:
 ```
           if get_model() == "acme":
                recommended_time = get_recommended_test_time(test)
                if recommended_time is not None:
                    create_newcase_cmd += " --walltime {}".format(recommended_time)
```

**Comment by jgfouca** [comment]
@sarich , at that point in the code, it expects the test names to be fully resolved, IE, TEST.GRID.COMPSET.MACHINE_COMPILER

**Comment by jgfouca** [comment]
My hunch is that you need to fully-resolve the test names before you give them to TestScheduler. There's a function in update_acme_tests that does that (get_full_test_names).

**Comment by jedwards4b** [comment]
It seems that only acme needs to have machine and compiler resolved by this point, in cesm we can allow it to pass machine and compiler as None to create_newcase and resolve them there.   I can add code to resolve these earlier.

**Comment by jedwards4b** [comment]
@sarich please try again, I now resolve machine and compiler earlier.

**Comment by sarich** [comment]
@jedwards4b, can you check that fix again? there's some commas missing in load_balance_submit.py, and after adding those the problem is stil there. The machine and compiler are getting passed to the TestScheduler ok, but CIME.utils.parse_test_name() doesn't have access to those values.

`./load_balancing_submit.py  --pesfile tests/pes.xml --res f19_g16 --compset X --force-purge`

````
Finished CREATE_NEWCASE for test PFS_I0.f19_g16.X in 0.016461 seconds (FAIL). [COMPLETED 1 of 3]
    Case dir: /homes/sarich/acme/scratch/PFS_I0.f19_g16.X.lbt
    Errors were:
        Test 'PFS_I0.f19_g16.X' failed in phase 'CREATE_NEWCASE' with exception 'ERROR: PFS_I0.f19_g16.X is not a full test name'
          File "/sandbox/sarich/cime/scripts/Tools/../../scripts/lib/CIME/test_scheduler.py", line 616, in _run_catch_exceptions
            return run(test)
          File "/sandbox/sarich/cime/scripts/Tools/../../scripts/lib/CIME/test_scheduler.py", line 441, in _create_newcase_phase
            recommended_time = get_recommended_test_time(test)
          File "/sandbox/sarich/cime/scripts/Tools/../../scripts/lib/update_acme_tests.py", line 317, in get_recommended_test_time
            expect(machine is not None, "{} is not a full test name".format(test_full_name))
          File "/sandbox/sarich/cime/scripts/Tools/../../scripts/lib/CIME/utils.py", line 49, in expect
            raise exc_type("{} {}".format(error_prefix, error_msg))
```

**Comment by jedwards4b** [comment]
Sorry - should be okay now.

**Comment by jedwards4b** [comment]
Working on updating documentation and tests.

**Comment by jedwards4b** [comment]
I have all the tests working again and @sarich I want to apologize.  You did some nice work here and I shouldn't have submitted this PR without making sure the tests you wrote still worked, they do now. 
 tests/load_balancing_test.py 
test_graph_models (__main__.LoadBalanceTests) ... ok
test_pulp (__main__.LoadBalanceTests) ... ok
test_read_and_write_json (__main__.LoadBalanceTests)
Solve from json file, writing to new json file, solve from new file ... ok
test_set_blocksize_atm (__main__.LoadBalanceTests) ... ok
test_solve_from_timing_dir (__main__.LoadBalanceTests) ... ok
test_use_atm_lnd (__main__.LoadBalanceTests)
Solve layout atm_lnd from json file ... ok
test_write_pes (__main__.LoadBalanceTests) ... ok
test_xcase_submit (__main__.LoadBalanceTests) ... ok

----------------------------------------------------------------------
Ran 8 tests in 22.924s

OK

**Comment by sarich** [comment]
Thanks for fixing everything up, I'll go over it again tomorrow morning. Is
the acme get recommended time working? I don't think it was ever fixed, it
may break all acme runs.

**Comment by jedwards4b** [comment]
If the problem was that you needed machine and compiler then it should be fixed, but I can't test it.

**Comment by sarich** [comment]
That's the problem, and it's not fixed. You should be able to reproduce the
problem by forcing through the 'if model == acme' branch

**Comment by jedwards4b** [comment]
@sarich my acme system seems to work fine.

**Comment by sarich** [comment]
Yes, you're right, ACME tests are working fine.

@jedwards4b,
Your change last night to include the full name of the test fixed the problem I was having. I didn't look at it yet because I thought it just a documentation and testing change, sorry.
	   
```
 -       test_names.append("PFS_I{}.{}.{}".format(i, res, compset))		
 +       test_names.append("PFS_I{}.{}.{}.{}_{}".format(i, res, compset, machine, compiler))
```



I'm having a problem with ./load_balance_solve.py finding the timing files. I'm looking into it, hopefully won't take long.
```
sarich@compute001:/sandbox/sarich/cime/tools/load_balancing_tool$ ./load_balancing_solve.py  --total-tasks 32
Traceback (most recent call last):
  File "./load_balancing_solve.py", line 318, in <module>
    _main_func(__doc__)
  File "./load_balancing_solve.py", line 313, in _main_func
    sys.exit(load_balancing_solve(test_id, test_root, timing_dir, blocksizes, total_tasks, layout, graph_models, print_models, pe_output, json_output, json_input))
  File "./load_balancing_solve.py", line 259, in load_balancing_solve
    timing_dir)
  File "./load_balancing_solve.py", line 123, in _locate_timing_files
    for fn in os.listdir(test_root):
TypeError: coercing to Unicode: need string or buffer, NoneType found

```
Thanks for your help on this, I'm very happy with the changes.

**Comment by sarich** [comment]
Sorry, pushed the wrong button and accidentally closed it.

**Comment by jedwards4b** [comment]
@sarich I think that I've fixed the issue in load_balance_solve.py

**Comment by sarich** [comment]
Yep, that did it, thanks. I'm happy with the PR now.

File: tools/load_balancing_tool/load_balancing_submit.py [hunk 3]
```diff
@@ -228,97 +222,94 @@ def load_balancing_submit(compset, res, pesfile, compiler, project, machine,
...
+    if compiler is None:
+        compiler = machobj.get_default_compiler()
+        print "compiler is {}".format(compiler)
+    if mpilib is None:
+        mpilib = machobj.get_default_MPIlib()
+
+
+    test_names = []
+    for i in xrange(len(pesize_list)):
+        test_names.append("PFS_I{}.{}.{}.{}_{}".format(i, res, compset, machine, compiler))
```
Comment (jgfouca) [inline tools/load_balancing_tool/load_balancing_submit.py]: This should use the standard API for creating a "full" test name: update_acme_tests.get_full_test_names()

File: tools/load_balancing_tool/tests/load_balancing_test.py [hunk 4]
```diff
@@ -232,35 +232,42 @@ def test_graph_models(self):
...
-            self.assertTrue(output.find("Timing jobs submitted") >= 0,
-                            "Expected 'Timing jobs submitted' in output")
-            
-            expected_dir = os.path.join(SCRIPT_DIR, "test_lbt_1", "timing")
+        test_root = MACHINE.get_value("CIME_OUTPUT_ROOT")
+        machine = MACHINE.get_machine_name()
+        compiler = MACHINE.get_default_compiler()
+
+        expected_dir = os.path.join(test_root,
+                                    "PFS_I0.f19_g16.X.{}_{}.test_lbt".format(machine,compiler),
```
Comment (jgfouca) [inline tools/load_balancing_tool/tests/load_balancing_test.py]: Use API for full test name

**Comment by jedwards4b** [comment]
Thanks @jgfouca  - it's already in CIME:utils so no need to move anything

**Comment by jgfouca** [comment]
@jedwards4b oh good, I must have been looking at an old branch

**Comment by rljacob** [comment]
Is the README duplicating a lot of the info in the users guide?  If so, the README should be shortened and just point people to the user's guide.

**Comment by jedwards4b** [comment]
@jgfouca I have another PR to submit as soon as this one is merged.  I will cleanup the README in that one.

**Pull request merged**